File: session-bg/plugins/fx/matrix.lua
local W, H = 0, 0
local rng = nil
local drops = {}
local shimmer = 0.0

local function target_drops(density)
  return math.max(1, math.floor(W * 0.22 * density))
end

local function spawn(above)
  if W == 0 or H == 0 then
    return
  end
  local len = math.max(3, math.floor(rng:range(4.0, math.max(H, 6) * 0.6)))
  local head
  if above then
    head = -rng:range(0.0, H)
  else
    head = rng:range(0.0, H)
  end
  local glyphs = {}
  for i = 1, len + 2 do
    glyphs[i] = sbg.glyphs.matrix[rng:below(#sbg.glyphs.matrix) + 1]
  end
  drops[#drops + 1] = {
    x = rng:below(W),
    head = head,
    len = len,
    speed = rng:range(4.0, 14.0),
    glyphs = glyphs,
  }
end

function init(ctx)
  W, H = ctx.w, ctx.h
  rng = sbg.rng(ctx.seed)
  drops = {}
  shimmer = 0.0
  for _ = 1, target_drops(ctx.density) do
    spawn(false)
  end
end

function step(dt, state)
  shimmer = shimmer + dt
  for _, d in ipairs(drops) do
    d.head = d.head + d.speed * dt
    if rng:chance(dt * 6.0) then
      local i = rng:below(#d.glyphs) + 1
      d.glyphs[i] = sbg.glyphs.matrix[rng:below(#sbg.glyphs.matrix) + 1]
    end
  end
  local kept = {}
  for _, d in ipairs(drops) do
    if d.head - d.len < H then
      kept[#kept + 1] = d
    end
  end
  drops = kept
  local density = state.params.density or 1.0
  local wanted = target_drops(density)
  while #drops < wanted do
    spawn(true)
  end
  if state.mod.burst > 0.2 and rng:chance(dt * 4.0 * state.mod.burst) then
    spawn(true)
  end
end

function render(fx, state)
  local head_col = { sbg.hex(0xd7ffe0) }
  local bright = { sbg.hex(0x9ece6a) }
  local dark = { sbg.hex(0x1f5a2a) }
  for _, d in ipairs(drops) do
    local head = math.floor(d.head)
    for i = 0, d.len - 1 do
      local y = head - i
      if y >= 0 and y < H then
        local t = i / d.len
        local r, g, b
        if i == 0 then
          r, g, b = head_col[1], head_col[2], head_col[3]
        else
          r, g, b = sbg.mix(bright[1], bright[2], bright[3], dark[1], dark[2], dark[3], t)
          local pulse = 0.85 + 0.15 * math.sin(shimmer * 3.0 + d.x)
          r, g, b = sbg.scale(r, g, b, (1.0 - t * 0.6) * pulse)
        end
        local ch = d.glyphs[(i % #d.glyphs) + 1]
        fx:put(d.x, y, ch, r, g, b)
      end
    end
  end
end

File: session-bg/plugins/src/effects/matrix.rs
use super::Effect;
use crate::frame::{hex, mix, scale, Glyph};
use crate::rng::Rng;

const GLYPHS: &[char] = &[
    'ｱ', 'ｲ', 'ｳ', 'ｴ', 'ｵ', 'ｶ', 'ｷ', 'ｸ', 'ｹ', 'ｺ', 'ｻ', 'ｼ', 'ｽ', 'ｾ', 'ｿ', 'ﾀ', 'ﾁ', 'ﾂ', 'ﾃ',
    'ﾄ', 'ﾅ', 'ﾆ', 'ﾈ', 'ﾉ', 'ﾊ', 'ﾋ', 'ﾌ', 'ﾍ', 'ﾎ', 'ﾏ', 'ﾐ', 'ﾑ', 'ﾒ', 'ﾓ', 'ﾔ', 'ﾕ', 'ﾖ', 'ﾗ',
    'ﾘ', 'ﾙ', 'ﾚ', 'ﾜ', 'ﾝ', '0', '1', '2', '3', '5', '7', '8', '9', 'Z', 'X', ':', '=', '*', '+',
    '<', '>',
];

struct Drop {
    x: u16,
    head: f32,
    len: u16,
    speed: f32,
    glyphs: Vec<char>,
}

pub struct Matrix {
    width: u16,
    height: u16,
    density: f32,
    drops: Vec<Drop>,
    shimmer: f32,
}

impl Matrix {
    pub fn new(density: f32) -> Self {
        Self {
            width: 0,
            height: 0,
            density,
            drops: Vec::new(),
            shimmer: 0.0,
        }
    }

    fn target_drops(&self) -> usize {
        ((f32::from(self.width) * 0.22 * self.density).round() as usize).max(1)
    }

    fn spawn(&mut self, rng: &mut Rng, above: bool) {
        if self.width == 0 || self.height == 0 {
            return;
        }
        let len = rng.range(4.0, f32::from(self.height).max(6.0) * 0.6) as u16;
        let head = if above {
            -rng.range(0.0, f32::from(self.height))
        } else {
            rng.range(0.0, f32::from(self.height))
        };
        let glyphs = (0..len + 2).map(|_| GLYPHS[rng.below(GLYPHS.len())]).collect();
        self.drops.push(Drop {
            x: rng.below(self.width as usize) as u16,
            head,
            len: len.max(3),
            speed: rng.range(4.0, 14.0),
            glyphs,
        });
    }
}

impl Effect for Matrix {
    fn resize(&mut self, width: u16, height: u16, rng: &mut Rng) {
        self.width = width;
        self.height = height;
        self.drops.clear();
        let n = self.target_drops();
        for _ in 0..n {
            self.spawn(rng, false);
        }
    }

    fn set_density(&mut self, density: f32) {
        self.density = density.clamp(0.1, 3.0);
    }

    fn step(&mut self, dt: f32, rng: &mut Rng) {
        self.shimmer += dt;
        let h = f32::from(self.height);
        for d in &mut self.drops {
            d.head += d.speed * dt;
            if rng.chance(dt * 6.0) {
                let i = rng.below(d.glyphs.len());
                d.glyphs[i] = GLYPHS[rng.below(GLYPHS.len())];
            }
        }
        self.drops.retain(|d| d.head - f32::from(d.len) < h);
        while self.drops.len() < self.target_drops() {
            self.spawn(rng, true);
        }
    }

    fn render(&self, out: &mut Vec<Glyph>) {
        let head_col = hex(0xd7ffe0);
        let bright = hex(0x9ece6a);
        let dark = hex(0x1f5a2a);
        for d in &self.drops {
            let head = d.head.floor() as i32;
            for i in 0..i32::from(d.len) {
                let y = head - i;
                if y < 0 || y >= i32::from(self.height) {
                    continue;
                }
                let t = i as f32 / f32::from(d.len);
                let rgb = if i == 0 {
                    head_col
                } else {
                    let base = mix(bright, dark, t);
                    let pulse = 0.85 + 0.15 * ((self.shimmer * 3.0 + f32::from(d.x)).sin());
                    scale(base, (1.0 - t * 0.6) * pulse)
                };
                out.push(Glyph {
                    x: d.x,
                    y: y as u16,
                    ch: d.glyphs[i as usize % d.glyphs.len()],
                    rgb,
                });
            }
        }
    }
}
